<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Adapted by Milan Malfait and Leo Fuhrhop" />


<title>Lab 6: Large Scale Inference</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>

<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HDDA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="intro.html">1. Introduction</a>
    </li>
    <li>
      <a href="svd.html">2. Singular Value Decomposition</a>
    </li>
    <li>
      <a href="prediction.html">3. Prediction with High Dimensional Predictors</a>
    </li>
    <li>
      <a href="sparseSvd.html">4. Sparse Singular Value Decomposition</a>
    </li>
    <li>
      <a href="lda.html">5. Linear Discriminant Analysis</a>
    </li>
    <li>
      <a href="hclust.html">6.1. Introduction to Clustering</a>
    </li>
    <li>
      <a href="https://sites.stat.washington.edu/people/raftery/Research/PDF/fraley1998.pdf">6.2. Paper Model-based Clustering</a>
    </li>
    <li>
      <a href="em.html">6.3. EM algorithm</a>
    </li>
    <li>
      <a href="lsi.html">7. Large Scale Inference</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Labs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Lab1-Intro-SVD.html">Lab 1: Intro &amp; SVD</a>
    </li>
    <li>
      <a href="Lab2-PCA.html">Lab 2: SVD - PCA</a>
    </li>
    <li>
      <a href="Lab3-Penalized-Regression.html">Lab 3: Prediction</a>
    </li>
    <li>
      <a href="Lab4-Sparse-PCA-LDA.html">Lab 4: Sparse PCA &amp; LDA</a>
    </li>
    <li>
      <a href="Lab5-Clustering.html">Lab 5: Clustering</a>
    </li>
    <li>
      <a href="Lab6-Large-Scale-Inference.html">Lab 6: LSI</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="https://github.com/statOmics/HDDA">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Lab 6: Large Scale Inference</h1>
<h3 class="subtitle">High Dimensional Data Analysis practicals</h3>
<h4 class="author">Adapted by Milan Malfait and Leo Fuhrhop</h4>
<h4 class="date">21 Feb 2022 <br/> (Last updated: 2025-11-26)</h4>

</div>


<div id="change-log" class="section level3 unnumbered">
<h3><a href="https://github.com/statOmics/HDDA/commits/master/Lab6-Large-Scale-Inference.Rmd">Change log</a></h3>
<hr />
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">## install packages with:</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co"># if (!requireNamespace(&quot;remotes&quot;, quietly = TRUE)) {</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#     install.packages(&quot;remotes&quot;)</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># remotes::install_github(&quot;statOmics/HDDAData&quot;)</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(HDDAData)</span></code></pre></div>
</div>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p><strong>In this lab session we will look at the following topics</strong></p>
<ul>
<li>FWER</li>
<li>FDR</li>
<li>Multiple testing problem in a real dataset</li>
</ul>
<div id="testing-hypotheses" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Testing many hypotheses</h2>
<p>To demonstrate the ideas we will be working with, we will simulate artificial data.
Note that since we are doing simulations, we can control everything and also know
exactly what the underlying “true” distribution is (sometimes also called the “ground truth”).
However, keep in mind that this is an unlikely representation of real-world data.</p>
<p>In particular, we will simulate data for multiple hypothesis tests where
<strong>the null hypothesis is always true</strong>. This means that for the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>-th test, if we let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{1i}</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{2i}</annotation></semantics></math> represent the means of the two populations of interest,
the null hypothesis for comparing the two means is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mrow><mn>0</mn><mi>i</mi></mrow></msub><mo>:</mo><msub><mi>μ</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><mo>=</mo><msub><mi>μ</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub></mrow><annotation encoding="application/x-tex">H_{0i} : \mu_{1i} = \mu_{2i}</annotation></semantics></math>. Let the alternative hypothesis of this test be
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><mo>:</mo><msub><mi>μ</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><mo>≠</mo><msub><mi>μ</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub></mrow><annotation encoding="application/x-tex">H_{1i} : \mu_{1i} \neq \mu_{2i}</annotation></semantics></math>, i.e., we perform a <em>two-tailed</em> test.
Suppose we have data collected from both populations, given by <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">X_{1i}</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">X_{2i}</annotation></semantics></math>
of size <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">n_{1i}</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">n_{2i}</annotation></semantics></math>, respectively, and assume that both populations
have the same known variance <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>i</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma_i^2</annotation></semantics></math>. Then we can test this hypothesis by using a
<strong>Z-test statistic</strong>, given by</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Z</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><msub><mover><mi>X</mi><mo accent="true">‾</mo></mover><mrow><mn>1</mn><mi>i</mi></mrow></msub><mo>−</mo><msub><mover><mi>X</mi><mo accent="true">‾</mo></mover><mrow><mn>2</mn><mi>i</mi></mrow></msub></mrow><mrow><msub><mi>σ</mi><mi>i</mi></msub><msqrt><mrow><mn>1</mn><mi>/</mi><msub><mi>n</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><mo>+</mo><mn>1</mn><mi>/</mi><msub><mi>n</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub></mrow></msqrt></mrow></mfrac><mspace width="0.167em"></mspace><mi>.</mi></mrow><annotation encoding="application/x-tex"> Z_i=\frac{\bar X_{1i}-\bar X_{2i}}{\sigma_i\sqrt{1/n_{1i}+1/n_{2i}}} \, .</annotation></semantics></math></p>
<p>Under the null hypothesis, the scores will be distributed as a <em>standard normal</em>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Z</mi><mi>i</mi></msub><mo>∼</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Z_i\sim N(0,1)</annotation></semantics></math>. The null hypothesis is rejected in favor of the alternative if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mi>i</mi></msub><mo>&lt;</mo><msub><mi>ϕ</mi><mrow><msub><mi>α</mi><mi>i</mi></msub><mi>/</mi><mn>2</mn></mrow></msub></mrow><annotation encoding="application/x-tex">z_i &lt; \phi_{\alpha_i/2}</annotation></semantics></math> or <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mi>i</mi></msub><mo>&gt;</mo><msub><mi>ϕ</mi><mrow><mn>1</mn><mo>−</mo><msub><mi>α</mi><mi>i</mi></msub><mi>/</mi><mn>2</mn></mrow></msub></mrow><annotation encoding="application/x-tex">z_i &gt; \phi_{1-\alpha_i/2}</annotation></semantics></math>, or equivalently if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>z</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>&gt;</mo><msub><mi>ϕ</mi><mrow><mn>1</mn><mo>−</mo><msub><mi>α</mi><mi>i</mi></msub><mi>/</mi><mn>2</mn></mrow></msub></mrow><annotation encoding="application/x-tex">|z_i| &gt; \phi_{1-\alpha_i/2}</annotation></semantics></math>, where <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϕ</mi><mrow><msub><mi>α</mi><mi>i</mi></msub><mi>/</mi><mn>2</mn></mrow></msub><annotation encoding="application/x-tex">\phi_{\alpha_i/2}</annotation></semantics></math> is the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>α</mi><mi>i</mi></msub><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">\alpha_i/2</annotation></semantics></math>th
quantile of the standard normal distribution.</p>
<p>In a multiple hypothesis testing setting, we will perform <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math> tests using the
same test statistic. If the <em>null</em> were true for all hypotheses, we would end up
with a sample <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mn>1</mn></msub><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>z</mi><mi>m</mi></msub></mrow><annotation encoding="application/x-tex">z_1,\ldots,z_m</annotation></semantics></math> from a standard normal distribution.</p>
<p>In this setting, we can only make one type of error: wrongly rejecting the null
hypothesis, i.e. a <strong>type 1 error</strong>. The probability of making this error is given
by</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>α</mi><mi>i</mi></msub><mo>=</mo><mtext mathvariant="normal">P</mtext><mo stretchy="false" form="prefix">{</mo><mrow><mspace width="0.333em"></mspace><mtext mathvariant="normal"> reject </mtext><mspace width="0.333em"></mspace></mrow><msub><mi>H</mi><mrow><mn>0</mn><mi>i</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>H</mi><mrow><mn>0</mn><mi>i</mi></mrow></msub><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex"> \alpha_i=\text{P}\{\text{ reject } H_{0i} | H_{0i}\} </annotation></semantics></math></p>
<p>(the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false" form="prefix">|</mo><msub><mi>H</mi><mrow><mn>0</mn><mi>i</mi></mrow></msub></mrow><annotation encoding="application/x-tex">| H_{0i}</annotation></semantics></math> part should be read as “given that the null hypothesis is true”).</p>
<p>If we now perform <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>m</mi><mn>0</mn></msub><annotation encoding="application/x-tex">m_0</annotation></semantics></math> such tests (using the 0 subscript to denote that they
are all hypotheses for which the <em>null</em> is true), we can summarise the possible
outcomes as in the table below:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Accept H_0</th>
<th align="left">Reject H_0</th>
<th align="left">Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">H_0 True</td>
<td align="left">U (True Negative)</td>
<td align="left">V (False Positive)</td>
<td align="left">m_0</td>
</tr>
</tbody>
</table>
<p>Here, <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>U</mi><annotation encoding="application/x-tex">U</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>V</mi><annotation encoding="application/x-tex">V</annotation></semantics></math> represent the total number of true negative and false positive
results we get, respectively, out of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>m</mi><mn>0</mn></msub><annotation encoding="application/x-tex">m_0</annotation></semantics></math> total tests.</p>
<p>As an example, let’s simulate the scenario above. In the code below, the <code>replicate</code>
function is used to repeat the same procedure a number of times.
Essentially, the following steps are performed :</p>
<ol style="list-style-type: decimal">
<li>Sample <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mo>=</mo><mn>50</mn></mrow><annotation encoding="application/x-tex">N = 50</annotation></semantics></math> observations for 2 hypothetical groups from normal distributions
with the same mean and known variance.</li>
<li>Calculate Z-test statistics based on the 2 group means.</li>
<li>Repeat steps 1 to 2 <code>m0</code> times.</li>
</ol>
<p>This mimics performing <code>m0</code> hypothesis tests on data for which we know the null
hypothesis is true.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Set parameters for the simulation</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">50</span> <span class="co"># samples per group</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>m0 <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of hypothesis tests</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>mu_1 <span class="ot">&lt;-</span> <span class="dv">3</span> <span class="co"># true mean group 1</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>mu_2 <span class="ot">&lt;-</span> <span class="dv">3</span> <span class="co"># true mean group 2</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>sigma <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="co"># known variance</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>denom <span class="ot">&lt;-</span> sigma <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">2</span> <span class="sc">/</span> N) <span class="co"># denominator for Z-scores</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>) <span class="co"># seed for reproducibility</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>z_stats <span class="ot">&lt;-</span> <span class="fu">replicate</span>(m0, {</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>  group1 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(N, <span class="at">mean =</span> mu_1, <span class="at">sd =</span> <span class="fu">sqrt</span>(sigma))</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>  group2 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(N, <span class="at">mean =</span> mu_2, <span class="at">sd =</span> <span class="fu">sqrt</span>(sigma))</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>  <span class="co"># Calculate Z-test statistic</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>  (<span class="fu">mean</span>(group2) <span class="sc">-</span> <span class="fu">mean</span>(group1)) <span class="sc">/</span> denom</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>})</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co"># Visualize Z-test statistics</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="fu">hist</span>(z_stats, <span class="at">breaks =</span> <span class="dv">50</span>, <span class="at">freq =</span> <span class="cn">FALSE</span>, <span class="at">main =</span> <span class="st">&quot;Z-test statistics under H_0&quot;</span>)</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co"># Overlay theoretical standard normal</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="fu">lines</span>(x <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>, <span class="at">length.out =</span> <span class="dv">100</span>), <span class="fu">dnorm</span>(x), <span class="at">col =</span> <span class="st">&quot;dodgerblue&quot;</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co"># Draw vertical lines at 2.5 and 97.5th percentiles</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.025</span>, <span class="fl">0.975</span>)), <span class="at">col =</span> <span class="st">&quot;firebrick&quot;</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span></code></pre></div>
<p><img src="Lab6-Large-Scale-Inference_files/figure-html/unnamed-chunk-2-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>We see that the Z-test statistics nicely follow a standard normal distribution.
The vertical dashed lines indicate the 2.5th and 97.5th percentiles of the standard
normal. The regions outside these lines indicate the Z-scores that we would call
significant if we used a cut-off of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.05</annotation></semantics></math> for a <em>two-tailed</em> test.
So, even though we simulated data under the null hypothesis, our Z-test still
returns “significant” results for a number of cases just by chance!</p>
<p>Let’s calculate the p-values for our hypothesis tests and see what the damage is.
To calculate the p-values, we use the <code>pnorm()</code> function in this case, which returns
the value of the standard normal CDF (i.e. <code>pnorm(x)</code> = <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mo>&lt;</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(Z &lt; x)</annotation></semantics></math>). Since we consider
a two-tailed test, we take the absolute values of the Z-scores and set the <code>lower.tail</code>
argument in <code>pnorm</code> to <code>FALSE</code> (by default it’s <code>TRUE</code>), so that we get
<code>pnorm(abs(x), lower.tail = FALSE)</code> = <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mo>&gt;</mo><mrow><mo stretchy="true" form="prefix">|</mo><mi>x</mi><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(Z &gt; |x|)</annotation></semantics></math> and multiply this value by 2.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>null_pvals <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> <span class="fu">pnorm</span>(<span class="fu">abs</span>(z_stats), <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">hist</span>(null_pvals, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.05</span>))</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> alpha, <span class="at">col =</span> <span class="st">&quot;firebrick&quot;</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span></code></pre></div>
<p><img src="Lab6-Large-Scale-Inference_files/figure-html/unnamed-chunk-3-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>called <span class="ot">&lt;-</span> (null_pvals <span class="sc">&lt;</span> alpha)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co"># V = number of false positives, in this case: all significant tests</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>(V <span class="ot">&lt;-</span> <span class="fu">sum</span>(called))</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; [1] 51</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="fu">mean</span>(called) <span class="co"># V / m0</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; [1] 0.051</span></span></code></pre></div>
<p>So we get 51 significant tests (false positives) out of a total of 1000, which is,
unsurprisingly, approximately equal to our significance cutoff <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>.
Note also that the p-values are uniformly distibuted under the null hypothesis by definition.</p>
<p>If we had carried out only a few tests (say 10) it would be very unlikely to observe
a false positive (on average: 0.05 * 10 = 0.5 false positives) at <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.05</annotation></semantics></math>,
but since now we’re carrying out so many, we’re almost guaranteed to get false positives.
This is what is known as the <strong>multiple hypothesis testing problem</strong>.</p>
<p>Note that in real-world data the quantities <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>U</mi><annotation encoding="application/x-tex">U</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>V</mi><annotation encoding="application/x-tex">V</annotation></semantics></math> are <em>unknown</em> (because
we don’t know the truth; if we did, we wouldn’t have to carry out any hypothesis
tests in the first place). However, by using simulated data, we do know the truth
and so we can explore these quantities.</p>
</div>
<div id="the-family-wise-error-rate" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> The family-wise error rate</h2>
<p>If we carry out many tests, we’re almost guaranteed to get type I errors, just by
chance. Therefore the type I error rate is no longer a relevant metric.
Instead, we consider the <strong>Familywise Error Rate (FWER)</strong>, given by</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">FWER</mtext><mo>=</mo><mtext mathvariant="normal">P</mtext><mo stretchy="false" form="prefix">{</mo><mi>V</mi><mo>&gt;</mo><mn>0</mn><mo stretchy="false" form="postfix">}</mo><mo>=</mo><mtext mathvariant="normal">P</mtext><mo stretchy="false" form="prefix">{</mo><mrow><mtext mathvariant="normal">rejecting at least one </mtext><mspace width="0.333em"></mspace></mrow><msub><mi>H</mi><mrow><mn>0</mn><mi>i</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>H</mi><mn>0</mn></msub><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">
\text{FWER}=\text{P}\{V &gt; 0\}
  = \text{P}\{\text{rejecting at least one } H_{0i}| H_0\}
</annotation></semantics></math></p>
<p>where <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>H</mi><mn>0</mn></msub><annotation encoding="application/x-tex">H_0</annotation></semantics></math> is the intersection of all partial nulls (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>H</mi><mrow><mn>0</mn><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">H_{0i}</annotation></semantics></math>) <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>m</mi><mn>0</mn></msub></mrow><annotation encoding="application/x-tex">i=1,\ldots,m_0</annotation></semantics></math>.
In general, we would prefer testing procedures that keep the FWER under control,
i.e. correct for the multiplicity problem.
So instead of choosing <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math> to control the probability of getting a false positive
in each test, we try to control the FWER, i.e. the probability of getting
<em>at least</em> one false positive in our <em>set</em> of hypothesis tests.</p>
<p>In the following exercises, we will explore this idea on some synthetic data. We will
simulate performing hypothesis tests for which the null distribution is always true
for different values of <code>m0</code>.</p>
<p>To simplify the code, we will directly sample the Z-test statistics from
a standard normal (instead of sampling the individual groups and then performing
the test on them, as shown above). We can also skip the p-value calculation and
compare our Z-test statistics directly to the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>α</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">(1 - \alpha) / 2</annotation></semantics></math>th quantile of the standard
normal distribution (using the <code>qnorm()</code> function).</p>
<div id="tasks" class="section level3 unnumbered">
<h3>Tasks</h3>
<div id="set-m0-5-and-generate-m0-z-test-statistics-by-sampling-from-the-standard-normal-distribution.-check-which-ones-are-significant-using-a-cutoff-alpha-0.05" class="section level4 unnumbered">
<h4>1. Set <code>m0 = 5</code> and generate <code>m0</code> Z-test statistics by sampling from the standard normal distribution. Check which ones are significant using a cutoff <code>alpha = 0.05</code></h4>
</div>
<div id="calculate-v-the-number-of-significant-tests." class="section level4 unnumbered">
<h4>2. Calculate <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>V</mi><annotation encoding="application/x-tex">V</annotation></semantics></math>, the number of significant tests.</h4>
</div>
<div id="check-whether-v-0-at-least-one-significant-and-store-the-result-in-a-variable." class="section level4 unnumbered">
<h4>3. Check whether <code>V &gt; 0</code> (at least one significant) and store the result in a variable.</h4>
</div>
<div id="repeat-steps-1-3-1000-times-and-keep-track-of-the-result-in-step-3-by-storing-it-in-a-vector." class="section level4 unnumbered">
<h4>4. Repeat steps 1-3 1000 times and keep track of the result in step 3 by storing it in a vector.</h4>
</div>
<div id="now-compute-the-fwer-as-the-proportion-of-times-v-was-greater-than-0." class="section level4 unnumbered">
<h4>5. Now compute the FWER as the proportion of times <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>V</mi><annotation encoding="application/x-tex">V</annotation></semantics></math> was greater than 0.</h4>
</div>
<div id="repeat-the-same-procedure-for-m0-50-and-m0-1000.-interpret-the-results." class="section level4 unnumbered">
<h4>6. Repeat the same procedure for <code>m0 = 50</code> and <code>m0 = 1000</code>. Interpret the results.</h4>
<details>
<summary>
Solution
</summary>
<p>To avoid repeated code, the simulations over the different values of <code>m_0</code> are combined into one code block.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Repeat 1000 times for m0 = 5, 50 and 1000</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of simulations</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>m0_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">50</span>, <span class="dv">1000</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co"># Initialize empty vector to store FWER results</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>fwer <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length =</span> <span class="fu">length</span>(m0_vec))</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="fu">names</span>(fwer) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;m0 =&quot;</span>, m0_vec)</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co"># Loop over m0 values and simulate test statistics</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(m0_vec)) {</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>  <span class="co"># Simulate B times, each time recording V &gt; 0</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>  at_least_one_signif <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>    null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0_vec[i], <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>) <span class="co"># simulate z-test</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>    significant_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(null_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>    <span class="fu">sum</span>(significant_tests) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>  })</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>  fwer[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(at_least_one_signif) <span class="co"># proportion of V &gt;= 1</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>}</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>fwer</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="co">#&gt;    m0 = 5   m0 = 50 m0 = 1000 </span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co">#&gt;     0.235     0.925     1.000</span></span></code></pre></div>
</details>
</div>
</div>
</div>
</div>
<div id="controlling-the-fwer" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Controlling the FWER</h1>
<p>We saw in the previous exercises that when conducting a large number of hypothesis
tests the FWER becomes unacceptably large. So what can we do to lower it?
The most straightforward way would be to just lower <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math> (so we get less false positives).</p>
<p>Under the null hypothesis (so the p-values are uniformly distributed), we can write the FWER
for <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math> independent tests as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">FWER</mtext><mo>=</mo><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mtext mathvariant="normal">at least one rejection</mtext><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mo>−</mo><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mtext mathvariant="normal">no rejections</mtext><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mo>−</mo><munderover><mo>∏</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>p</mi><mi>i</mi></msub><mo>&gt;</mo><mi>α</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mo>−</mo><munderover><mo>∏</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>α</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mo>−</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>α</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>m</mi></msup><mspace width="0.167em"></mspace><mo>,</mo></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
\text{FWER} = P(\text{at least one rejection}) &amp;= 1 - P(\text{no rejections}) \\
  &amp;= 1 - \prod_{i=1}^{m} P(p_i &gt; \alpha) \\
  &amp;= 1 - \prod_{i=1}^{m} (1 - \alpha) \\
  &amp;= 1 - (1 - \alpha)^{m} \, ,
\end{align}
</annotation></semantics></math></p>
<p>where <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mi>i</mi></msub><annotation encoding="application/x-tex">p_i</annotation></semantics></math> is the p-value for the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>-th test.</p>
<p>Šidák’s correction solves the above equation at a given FWER level (e.g., 0.05)
for <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math> to control the FWER using</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>1</mn><mo>−</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mtext mathvariant="normal">FWER</mtext><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mi>/</mi><mi>m</mi></mrow></msup><mspace width="0.167em"></mspace><mi>.</mi></mrow><annotation encoding="application/x-tex"> \alpha = 1 - (1 - \text{FWER})^{1/m} \, .</annotation></semantics></math></p>
<p>Without assuming independent tests, the Bonferroni correction uses that</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">FWER</mtext></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mtext mathvariant="normal">at least one rejection</mtext><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mo>∪</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></msubsup><mo stretchy="false" form="prefix">{</mo><msub><mi>p</mi><mi>i</mi></msub><mo>≤</mo><mi>α</mi><mo stretchy="false" form="postfix">}</mo><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>≤</mo><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>p</mi><mi>i</mi></msub><mo>≤</mo><mi>α</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>m</mi><mi>α</mi><mspace width="0.167em"></mspace><mo>,</mo></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
\text{FWER} &amp;= P(\text{at least one rejection}) \\
  &amp;= P(\cup_{i=1}^m \{ p_i \le \alpha \}) \\
  &amp;\le \sum_{i=1}^{m} P(p_i \le \alpha) \\
  &amp;= m \alpha \, ,
\end{align}
</annotation></semantics></math></p>
<p>where the inequality follows from the union bound, and solves the equation
at a given FWER level (e.g., 0.05) for <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math> to control the FWER with</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mfrac><mtext mathvariant="normal">FWER</mtext><mi>m</mi></mfrac><mspace width="0.167em"></mspace><mi>.</mi></mrow><annotation encoding="application/x-tex"> \alpha = \frac{\text{FWER}}{m} \, .</annotation></semantics></math></p>
<div id="tasks-1" class="section level3 unnumbered">
<h3>Tasks</h3>
<div id="determine-the-alpha-for-šidáks-correction-at-m0-5-m0-50-and-m0-1000.-use-these-values-to-repeat-the-simulation-from-the-previous-exercise-and-confirm-that-the-fwer-is-controlled-at-0.05." class="section level4 unnumbered">
<h4>1. Determine the <code>alpha</code> for Šidák’s correction at <code>m0 = 5</code>, <code>m0 = 50</code> and <code>m0 = 1000</code>. Use these values to repeat the simulation from the previous exercise, and confirm that the FWER is controlled at 0.05.</h4>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Repeat 1000 times for m0 = 5, 50 and 1000</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of simulations</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>m0_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">50</span>, <span class="dv">1000</span>)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>alpha_vec <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> (<span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.05</span>)<span class="sc">^</span>(<span class="dv">1</span> <span class="sc">/</span> m0_vec)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>alpha_vec</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; [1] 1.020622e-02 1.025340e-03 5.129198e-05</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co"># Initialize empty vector to store FWER results</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>fwer <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length =</span> <span class="fu">length</span>(m0_vec))</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="fu">names</span>(fwer) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;m0 =&quot;</span>, m0_vec)</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co"># Loop over m0 values and simulate test statistics</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(m0_vec)) {</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  <span class="co"># Simulate B times, each time recording V &gt; 0</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>  at_least_one_signif <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>    null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0_vec[i], <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>) <span class="co"># simulate z-test</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>    significant_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(null_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha_vec[i] <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>    <span class="fu">sum</span>(significant_tests) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a>  })</span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>  fwer[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(at_least_one_signif) <span class="co"># proportion of V &gt;= 1</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a>}</span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a>fwer</span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt;    m0 = 5   m0 = 50 m0 = 1000 </span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt;     0.040     0.053     0.045</span></span></code></pre></div>
</details>
</div>
<div id="determine-the-alpha-for-the-bonferroni-correction-and-repeat-the-simulation-one-more-time.-how-do-the-results-compare-to-šidáks-correction" class="section level4 unnumbered">
<h4>2. Determine the <code>alpha</code> for the Bonferroni correction and repeat the simulation one more time. How do the results compare to Šidák’s correction?</h4>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Repeat 1000 times for m0 = 5, 50 and 1000</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of simulations</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>m0_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">50</span>, <span class="dv">1000</span>)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>alpha_vec <span class="ot">&lt;-</span> <span class="fl">0.05</span> <span class="sc">/</span> m0_vec</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>alpha_vec</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt; [1] 1e-02 1e-03 5e-05</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co"># Initialize empty vector to store FWER results</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>fwer <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length =</span> <span class="fu">length</span>(m0_vec))</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="fu">names</span>(fwer) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;m0 =&quot;</span>, m0_vec)</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co"># Loop over m0 values and simulate test statistics</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(m0_vec)) {</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>  <span class="co"># Simulate B times, each time recording V &gt; 0</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>  at_least_one_signif <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>    null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0_vec[i], <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>) <span class="co"># simulate z-test</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>    significant_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(null_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha_vec[i] <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>    <span class="fu">sum</span>(significant_tests) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>  })</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>  fwer[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(at_least_one_signif) <span class="co"># proportion of V &gt;= 1</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>}</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>fwer</span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a><span class="co">#&gt;    m0 = 5   m0 = 50 m0 = 1000 </span></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="co">#&gt;     0.039     0.051     0.045</span></span></code></pre></div>
</details>
<p>In general, controlling the FWER at levels such as 0.05 with Šidák’s correction
or the Bonferroni correction is very conservative, i.e., if there were true alternatives
(which we did not consider in the simulations), we might miss them.</p>
</div>
</div>
</div>
<div id="the-false-discovery-rate-fdr" class="section level1" number="3">
<h1><span class="header-section-number">3</span> The False Discovery Rate (FDR)</h1>
<p>Now let’s consider a situation where we have both true null and true alternative
cases. In this situation we are prone to two types of errors: false positives
(type 1, as in the previous case) and false negatives (type 2, i.e. accepting
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>H</mi><mn>0</mn></msub><annotation encoding="application/x-tex">H_0</annotation></semantics></math> while it’s actually false).</p>
<p>We can then extend the table from <a href="#testing-hypotheses">before</a>:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Accept H_0</th>
<th align="left">Reject H_0</th>
<th align="left">Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">H_0 True</td>
<td align="left">U (True Negative)</td>
<td align="left">V (False Positive)</td>
<td align="left">m_0</td>
</tr>
<tr class="even">
<td align="left">H_1 True</td>
<td align="left">T (False Negative)</td>
<td align="left">S (True Positive)</td>
<td align="left">m_1</td>
</tr>
<tr class="odd">
<td align="left">Total</td>
<td align="left">W</td>
<td align="left">R</td>
<td align="left">m</td>
</tr>
</tbody>
</table>
<p>In addition to the FWER, another important quantity in multiple hypothesis
testing is the <strong>False Discovery Rate (FDR)</strong>:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">FDR</mtext><mo>=</mo><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><mfrac><mi>V</mi><mi>R</mi></mfrac><mo stretchy="true" form="postfix">]</mo></mrow><mspace width="0.167em"></mspace><mo>,</mo></mrow><annotation encoding="application/x-tex">
\text{FDR} = E[\frac{V}{R}] \, ,
</annotation></semantics></math></p>
<p>which is the expected proportion of false positives among all positive tests.
The ratio <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mi>/</mi><mi>R</mi></mrow><annotation encoding="application/x-tex">V/R</annotation></semantics></math> is also called the False Discovery Proportion (FDP).</p>
<p>However, the only quantities from the table above that are observable
in real data are <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>W</mi><annotation encoding="application/x-tex">W</annotation></semantics></math>, <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>. So to illustrate the concept, we will again
make use of simulated data.</p>
<p>In the code below, 1.000 tests are simulated, for which 90% come from cases
where the null is true and the other 10% for which the alternative is true.
We assume the <em>effect size</em> (difference between the 2 groups) to be equal to 3
under the alternative, so that we can sample the test statistics from a normal
distribution with mean 3.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># total number of hypotheses</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>p0 <span class="ot">&lt;-</span> <span class="fl">0.9</span> <span class="co"># 90% of cases are true null</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>m0 <span class="ot">&lt;-</span> <span class="fu">round</span>(p0 <span class="sc">*</span> m)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="fu">round</span>((<span class="dv">1</span> <span class="sc">-</span> p0) <span class="sc">*</span> m)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0)</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>alt_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m1, <span class="at">mean =</span> <span class="dv">3</span>)</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>null_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(null_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>alt_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(alt_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>(U <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span>null_tests)) <span class="co"># true negatives</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; [1] 854</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>(V <span class="ot">&lt;-</span> <span class="fu">sum</span>(null_tests)) <span class="co"># false positives</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt; [1] 46</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>(S <span class="ot">&lt;-</span> <span class="fu">sum</span>(alt_tests)) <span class="co"># true positives</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co">#&gt; [1] 84</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a>(T <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span>alt_tests)) <span class="co"># false negatives</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co">#&gt; [1] 16</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a>R <span class="ot">&lt;-</span> V <span class="sc">+</span> S <span class="co"># total number of positives</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a>(FDP <span class="ot">&lt;-</span> V <span class="sc">/</span> R)</span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="co">#&gt; [1] 0.3538462</span></span></code></pre></div>
<p>We see that 35.38% of the positive cases are actually false positives.
To get an idea of the FDR, the expected FDP, we will repeat this procedure a
number of times in the following exercises.</p>
<div id="tasks-2" class="section level3 unnumbered">
<h3>Tasks</h3>
<div id="set-m0-90-and-m1-10-and-generate-m-100-test-statistics-m0-of-them-from-the-standard-normal-distribution-n0-1-null-distribution-and-m1-of-them-from-n3-1-the-alternative-distribution." class="section level4 unnumbered">
<h4>1. Set <code>m0 = 90</code> and <code>m1 = 10</code> and generate <code>m = 100</code> test statistics, <code>m0</code> of them from the standard normal distribution <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">N(0, 1)</annotation></semantics></math> (null distribution) and m1 of them from <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>3</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">N(3, 1)</annotation></semantics></math> (the alternative distribution).</h4>
</div>
<div id="test-for-significance-at-alpha-0.05-by-comparing-the-test-statistics-to-the-standard-normal-for-both-sets-of-tests." class="section level4 unnumbered">
<h4>2. Test for significance at <code>alpha = 0.05</code> by comparing the test statistics to the standard normal, for both sets of tests.</h4>
</div>
<div id="calculate-the-fdp." class="section level4 unnumbered">
<h4>3. Calculate the FDP.</h4>
</div>
<div id="repeat-steps-1-3-1000-times-and-keep-track-of-the-fdp-for-each-iteration.-then-estimate-the-fdr-as-the-mean-of-the-fdps.-interpret-the-results." class="section level4 unnumbered">
<h4>4. Repeat steps 1-3 1000 times and keep track of the FDP for each iteration. Then estimate the FDR as the mean of the FDPs. Interpret the results.</h4>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of simulations</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>m0 <span class="ot">&lt;-</span> <span class="dv">90</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co"># Simulate B times, each time recording V &gt; 0</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>FDP <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>  null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0, <span class="at">mean =</span> <span class="dv">0</span>)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>  alt_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m1, <span class="at">mean =</span> <span class="dv">3</span>)</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>  null_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(null_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  alt_tests <span class="ot">&lt;-</span> <span class="fu">abs</span>(alt_stats) <span class="sc">&gt;</span> <span class="fu">qnorm</span>(<span class="dv">1</span> <span class="sc">-</span> alpha <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>  V <span class="ot">&lt;-</span> <span class="fu">sum</span>(null_tests) <span class="co"># false positives</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">sum</span>(alt_tests) <span class="co"># true positives</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>  R <span class="ot">&lt;-</span> V <span class="sc">+</span> S</span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>  <span class="co"># Return V / R</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>  <span class="cf">if</span> (R <span class="sc">==</span> <span class="dv">0</span>) <span class="dv">0</span> <span class="cf">else</span> V <span class="sc">/</span> R</span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>})</span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>(FDR <span class="ot">&lt;-</span> <span class="fu">mean</span>(FDP))</span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#&gt; [1] 0.3304283</span></span></code></pre></div>
<p>The FDR tells us the expected proportion of positive tests that will be false
positives.</p>
</details>
<p>As with the FWER, we can now think of methods to control the FDR. A very popular
method that is widely used in large-scale inference for high-throughput
sequencing data is the <strong>Benjamini-Hochberg correction</strong>. It <em>adjusts</em> the
p-values such that the FDR is controlled at a certain level. Note that this can
equivalently be formulated as adjusting the significance cut-off <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>, similarly
to how we defined the FWER-controlling techniques.</p>
</div>
</div>
<div id="tasks-3" class="section level3 unnumbered">
<h3>Tasks</h3>
<div id="adjust-the-simulation-code-from-the-previous-task-to-estimate-the-fdr-under-the-benjamini-hochberg-correction." class="section level4 unnumbered">
<h4>1. Adjust the simulation code from the previous task to estimate the FDR under the Benjamini-Hochberg correction.</h4>
<p>Hint: Calculate p-values of the simulated null and alternative test statistics. Concatenate all p-values into a single vector and apply the Benjamini-Hochberg correction using <code>p.adjust</code> with <code>method = "BH"</code>. Compare the resulting adjusted p-values to <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math> to estimate the FDP and FDR.</p>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of simulations</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>m0 <span class="ot">&lt;-</span> <span class="dv">90</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># Simulate B times, each time recording V &gt; 0</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>FDP <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>  null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0, <span class="at">mean =</span> <span class="dv">0</span>)</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  alt_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m1, <span class="at">mean =</span> <span class="dv">3</span>)</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  p_null <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">pnorm</span>(<span class="sc">-</span><span class="fu">abs</span>(null_stats)))</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  p_alt <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">pnorm</span>(<span class="sc">-</span><span class="fu">abs</span>(alt_stats)))</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>  p_adj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(<span class="fu">c</span>(p_null, p_alt), <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>  V <span class="ot">&lt;-</span> <span class="fu">sum</span>(p_adj[<span class="dv">1</span><span class="sc">:</span>m0] <span class="sc">&lt;=</span> alpha) <span class="co"># false positives</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">sum</span>(p_adj[(m0 <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">:</span>(m0 <span class="sc">+</span> m1)] <span class="sc">&lt;=</span> alpha) <span class="co"># true positives</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>  R <span class="ot">&lt;-</span> V <span class="sc">+</span> S</span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>  <span class="co"># Return V / R</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>  <span class="cf">if</span> (R <span class="sc">==</span> <span class="dv">0</span>) <span class="dv">0</span> <span class="cf">else</span> V <span class="sc">/</span> R</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>})</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>(FDR <span class="ot">&lt;-</span> <span class="fu">mean</span>(FDP))</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="co">#&gt; [1] 0.04071404</span></span></code></pre></div>
</details>
</div>
<div id="repeat-the-simulation-using-method-bonferroni.-compare-the-resulting-fdr-and-interpret." class="section level4 unnumbered">
<h4>2. Repeat the simulation using <code>method = "bonferroni"</code>. Compare the resulting FDR and interpret.</h4>
<p>Note: The Bonferroni correction can either be formulated as adjusting the significance cut-off to <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mi>/</mi><mi>m</mi></mrow><annotation encoding="application/x-tex">\alpha / m</annotation></semantics></math> or, equivalently, as a p-value correction that multiplies each p-value by <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>.</p>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987654321</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of simulations</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>m0 <span class="ot">&lt;-</span> <span class="dv">90</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co"># Simulate B times, each time recording V &gt; 0</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>FDP <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>  null_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m0, <span class="at">mean =</span> <span class="dv">0</span>)</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>  alt_stats <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(m1, <span class="at">mean =</span> <span class="dv">3</span>)</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  p_null <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">pnorm</span>(<span class="sc">-</span><span class="fu">abs</span>(null_stats)))</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  p_alt <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">pnorm</span>(<span class="sc">-</span><span class="fu">abs</span>(alt_stats)))</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>  p_adj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(<span class="fu">c</span>(p_null, p_alt), <span class="at">method =</span> <span class="st">&quot;bonferroni&quot;</span>)</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>  V <span class="ot">&lt;-</span> <span class="fu">sum</span>(p_adj[<span class="dv">1</span><span class="sc">:</span>m0] <span class="sc">&lt;=</span> alpha) <span class="co"># false positives</span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">sum</span>(p_adj[(m0 <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">:</span>(m0 <span class="sc">+</span> m1)] <span class="sc">&lt;=</span> alpha) <span class="co"># true positives</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>  R <span class="ot">&lt;-</span> V <span class="sc">+</span> S</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  <span class="co"># Return V / R</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  <span class="cf">if</span> (R <span class="sc">==</span> <span class="dv">0</span>) <span class="dv">0</span> <span class="cf">else</span> V <span class="sc">/</span> R</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>})</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>(FDR <span class="ot">&lt;-</span> <span class="fu">mean</span>(FDP))</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="co">#&gt; [1] 0.0106</span></span></code></pre></div>
<p><strong>Interpretation</strong>: Both the Bonferroni correction and the Benjamini-Hochberg correction control the FDR at the desired level <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>.
However, the Bonferroni correction is more conservative, resulting in an FDR that is much lower than <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>. This means that the Bonferroni method results in a larger number of false negatives and could possibly lead to us missing interesting discoveries.</p>
</details>
</div>
</div>
</div>
<div id="multiple-testing-on-the-data-by-alon-et-al.-1999" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Multiple testing on the data by Alon <em>et al.</em> (1999)</h1>
<p>We will take another look at the dataset by <span class="citation">Alon et al. (1999)</span> on gene expression
levels in 40 tumour and 22 normal colon tissue samples.
We previously used this data in <a href="./Lab4-Sparse-PCA-LDA.html">Lab 4</a>.
This time, we are interested in finding genes that are expressed <em>significantly</em>
different between the tumor and normal samples. In other words, we will perform
hypothesis tests for each of the <strong>2000 genes</strong>. This is clearly a multiple
testing problem.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;Alon1999&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">str</span>(Alon1999[, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]) <span class="co"># first 10 columns</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#&gt; &#39;data.frame&#39;:    62 obs. of  10 variables:</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt;  $ Y : chr  &quot;t&quot; &quot;n&quot; &quot;t&quot; &quot;n&quot; ...</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt;  $ X1: num  8589 9164 3826 6246 3230 ...</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt;  $ X2: num  5468 6720 6970 7824 3694 ...</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt;  $ X3: num  4263 4883 5370 5956 3401 ...</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt;  $ X4: num  4065 3718 4706 3976 3464 ...</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt;  $ X5: num  1998 2015 1167 2003 2181 ...</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt;  $ X6: num  5282 5570 1572 2131 2923 ...</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt;  $ X7: num  2170 3849 1325 1531 2069 ...</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt;  $ X8: num  2773 2793 1472 1715 2949 ...</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt;  $ X9: num  7526 7018 3297 3870 3303 ...</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="fu">table</span>(Alon1999<span class="sc">$</span>Y)  <span class="co"># contingency table of tissue types</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="co">#&gt;  n  t </span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="co">#&gt; 22 40</span></span></code></pre></div>
<p>As a reminder, the data consists of gene expression levels in 40 tumor and 22
normal colon tissue samples. The expression on 6500 human genes were measured
using the Affymetrix oligonucleotide array.
As in <span class="citation">Alon et al. (1999)</span>, we use the <em>2000 genes with the highest minimal intensity</em>
across the samples.</p>
<p>The dataset contains one variable named <code>Y</code> with the values <code>t</code> and <code>n</code>.
This variable indicates whether the sample came from tumourous (<code>t</code>) or
normal (<code>n</code>) tissue.</p>
<div id="tasks-4" class="section level3 unnumbered">
<h3>Tasks</h3>
<div id="perform-two-sample-t-tests-for-each-gene-comparing-the-tumor-and-normal-tissue-groups.-store-the-p-values-and-test-statistics." class="section level4 unnumbered">
<h4>1. Perform two-sample t-tests for each gene, comparing the tumor and normal tissue groups. Store the p-values and test statistics.</h4>
<p>Hint: The <code>t.test()</code> function computes a two-sample t-test with <em>unequal variances</em> (<a href="https://en.wikipedia.org/wiki/Welch%27s_t-test">Welch’s t-test</a>) by default. The return object contains the elements <code>p.value</code> and <code>statistic</code>.</p>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Split gene data and grouping variable and convert to matrix</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>gene_data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(Alon1999[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>group <span class="ot">&lt;-</span> Alon1999<span class="sc">$</span>Y</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co"># Use `apply` to perform t-tests over all columns</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>t_test_results <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(gene_data, <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>  t_test <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x <span class="sc">~</span> group)</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>  p_val <span class="ot">&lt;-</span> t_test<span class="sc">$</span>p.value</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>  stat <span class="ot">&lt;-</span> t_test<span class="sc">$</span>statistic</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>  <span class="fu">c</span>(stat, <span class="st">&quot;p_val&quot;</span> <span class="ot">=</span> p_val)</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>}))</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="fu">head</span>(t_test_results)</span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="co">#&gt;             t      p_val</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="co">#&gt; X1 -1.6728613 0.10068961</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="co">#&gt; X2 -1.4319967 0.15787615</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="co">#&gt; X3 -2.1292624 0.03758201</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="co">#&gt; X4 -0.9455604 0.34869002</span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="co">#&gt; X5 -1.7852252 0.07953894</span></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="co">#&gt; X6 -1.4214810 0.16143092</span></span></code></pre></div>
</details>
</div>
<div id="plot-a-histogram-of-the-p-values-and-interpret." class="section level4 unnumbered">
<h4>2. Plot a histogram of the p-values and interpret.</h4>
<details>
<summary>
Solution
</summary>
<p>Setting <code>breaks</code> from 0 to 1 with a width of 0.05 makes sense for p-values because
the first bar will then represent the number of p-values smaller than 0.05 (which
we often use as cutoff).</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>p_vals <span class="ot">&lt;-</span> t_test_results[, <span class="st">&quot;p_val&quot;</span>]</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">hist</span>(</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>  p_vals,</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.05</span>), <span class="at">main =</span> <span class="st">&quot;&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;p-value&quot;</span>,</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>  <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">500</span>)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="Lab6-Large-Scale-Inference_files/figure-html/p_val-hist-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><strong>Interpretation</strong>: The histogram appears close to a uniform distribution for the larger p-values, but has a lot more small p-values than would be expected under a uniform distribution.
Recall that if the null hypothesis (the gene expression is the same for tumor and normal tissue) were to hold for all genes, the p-values would be uniformly distributed on <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[0, 1]</annotation></semantics></math>.</p>
</details>
</div>
<div id="how-many-discoveries-significant-genes-do-you-find-when-using-alpha-0.05" class="section level4 unnumbered">
<h4>3. How many discoveries / significant genes do you find when using <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.05</annotation></semantics></math>?</h4>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">sum</span>(p_vals <span class="sc">&lt;=</span> alpha)</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt; [1] 478</span></span></code></pre></div>
</details>
</div>
<div id="control-the-fwer-at-5-using-šidáks-correction.-how-many-genes-are-significant-after-the-correction" class="section level4 unnumbered">
<h4>4. Control the FWER at 5% using Šidák’s correction. How many genes are significant after the correction?</h4>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">nrow</span>(t_test_results)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>fwer <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>adj_alpha <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> (<span class="dv">1</span> <span class="sc">-</span> fwer)<span class="sc">^</span>(<span class="dv">1</span> <span class="sc">/</span> m)</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co"># Number of significant discoveries</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="fu">sum</span>(p_vals <span class="sc">&lt;=</span> adj_alpha)</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#&gt; [1] 11</span></span></code></pre></div>
<p>By rearranging the equation <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>1</mn><mo>−</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mtext mathvariant="normal">FWER</mtext><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mi>/</mi><mi>m</mi></mrow></msup><mspace width="0.167em"></mspace><mo>,</mo></mrow><annotation encoding="application/x-tex"> \alpha = 1 - (1 - \text{FWER})^{1/m} \, ,</annotation></semantics></math>
we can equivalently adjust the p-values via
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>m</mi></msup></mrow><annotation encoding="application/x-tex">1 - (1 - p)^m</annotation></semantics></math>
and compare them to the cut-off 5%:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>sidak_p_vals <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> (<span class="dv">1</span> <span class="sc">-</span> p_vals)<span class="sc">^</span>m</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co"># Number of significant discoveries</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="fu">sum</span>(sidak_p_vals <span class="sc">&lt;=</span> fwer)</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="co">#&gt; [1] 11</span></span></code></pre></div>
</details>
</div>
<div id="control-the-fwer-at-5-using-the-bonferroni-correction.-how-many-genes-are-significant-after-the-correction" class="section level4 unnumbered">
<h4>5. Control the FWER at 5% using the Bonferroni correction. How many genes are significant after the correction?</h4>
<details>
<summary>
Solution
</summary>
<p>By adjusting <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>adj_alpha <span class="ot">&lt;-</span> fwer <span class="sc">/</span> m</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="co"># Number of significant discoveries</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="fu">sum</span>(p_vals <span class="sc">&lt;=</span> adj_alpha)</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co">#&gt; [1] 11</span></span></code></pre></div>
<p>Or equivalently, by adjusting the p-values:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>bonferroni_p_vals <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(p_vals, <span class="at">method =</span> <span class="st">&quot;bonferroni&quot;</span>)</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="co"># Number of significant discoveries</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="fu">sum</span>(bonferroni_p_vals <span class="sc">&lt;=</span> fwer)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="co">#&gt; [1] 11</span></span></code></pre></div>
</details>
</div>
<div id="control-the-fdr-at-5-using-the-benjamini-hochberg-method.-how-many-genes-are-significant-after-the-correction" class="section level4 unnumbered">
<h4>6. Control the FDR at 5% using the Benjamini-Hochberg method. How many genes are significant after the correction?</h4>
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>bh_p_vals <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(p_vals, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="fu">sum</span>(bh_p_vals <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="co">#&gt; [1] 190</span></span></code></pre></div>
</details>
</div>
<div id="plot-the-adjusted-p-values-from-the-benjamini-hochberg-method-against-the-original-p-values.-what-is-the-effect-of-the-adjustment" class="section level4 unnumbered">
<h4>7. Plot the adjusted p-values from the Benjamini-Hochberg method against the original p-values. What is the effect of the adjustment?</h4>
<p>Bonus: Compute the adjusted p-values based on Šidák’s correction and the Bonferroni correction. Add them to the plot and compare.</p>
<details>
<summary>
Solution
</summary>
<p>Visualize the adjusted p-values for all three methods in the same figure:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;#9d05f4&quot;</span>, <span class="st">&quot;#ecd033&quot;</span>, <span class="st">&quot;#01d9d9&quot;</span>)</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>  p_vals[<span class="fu">order</span>(p_vals)], bh_p_vals[<span class="fu">order</span>(p_vals)],</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">xlab =</span> <span class="st">&quot;p-value&quot;</span>,</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;Adjusted p-value&quot;</span>, <span class="at">col =</span> cols[<span class="dv">1</span>]</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>)</span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="fu">points</span>(</span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>  p_vals[<span class="fu">order</span>(p_vals)], bonferroni_p_vals[<span class="fu">order</span>(p_vals)],</span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">col =</span> cols[<span class="dv">2</span>]</span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>)</span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a><span class="fu">points</span>(</span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a>  p_vals[<span class="fu">order</span>(p_vals)], sidak_p_vals[<span class="fu">order</span>(p_vals)],</span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">col =</span> cols[<span class="dv">3</span>]</span>
<span id="cb21-15"><a href="#cb21-15" tabindex="-1"></a>)</span>
<span id="cb21-16"><a href="#cb21-16" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>)</span>
<span id="cb21-17"><a href="#cb21-17" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>,</span>
<span id="cb21-18"><a href="#cb21-18" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Benjamini-Hochberg&quot;</span>, <span class="st">&quot;Bonferroni&quot;</span>, <span class="st">&quot;Sidak&quot;</span>),</span>
<span id="cb21-19"><a href="#cb21-19" tabindex="-1"></a>  <span class="at">col =</span> cols, <span class="at">pch =</span> <span class="dv">19</span></span>
<span id="cb21-20"><a href="#cb21-20" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="Lab6-Large-Scale-Inference_files/figure-html/fdr-adjust-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><strong>Interpretation</strong>: All three methods adjust the p-values upwards to account for multiple testing. The adjustment is more pronounced for small p-values, while for higher p-values, the curves of the adjusted p-values flatten out.
The Benjamini-Hochberg method at a FDR of 5% is less conservative compared to the other two methods. Both Šidák’s correction and the Bonferroni correction at an FWER of 5% adjust the majority of p-values to exactly 1 and only a small proportion of p-values that are originally close to zero remains smaller than 1.</p>
</details>
</div>
</div>
</div>
<div id="further-reading" class="section level1 unnumbered">
<h1>Further reading</h1>
<ul>
<li>Chapter 6 of the <a href="http://genomicsclass.github.io/book/">Biomedical Data Science</a> book</li>
</ul>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1>Session info</h1>
<details>
<summary>
Session info
</summary>
<pre><code>#&gt; [1] &quot;2025-11-26 17:27:57 CET&quot;
#&gt; ─ Session info ───────────────────────────────────────────────────────────────
#&gt;  setting  value
#&gt;  version  R version 4.5.2 (2025-10-31)
#&gt;  os       macOS Sequoia 15.6
#&gt;  system   aarch64, darwin20
#&gt;  ui       X11
#&gt;  language (EN)
#&gt;  collate  en_US.UTF-8
#&gt;  ctype    en_US.UTF-8
#&gt;  tz       Europe/Brussels
#&gt;  date     2025-11-26
#&gt;  pandoc   3.6.3 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
#&gt;  quarto   1.7.32 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────
#&gt;  package     * version date (UTC) lib source
#&gt;  bookdown      0.45    2025-10-03 [1] CRAN (R 4.5.0)
#&gt;  bslib         0.9.0   2025-01-30 [1] CRAN (R 4.5.0)
#&gt;  cachem        1.1.0   2024-05-16 [1] CRAN (R 4.5.0)
#&gt;  cli           3.6.5   2025-04-23 [1] CRAN (R 4.5.0)
#&gt;  digest        0.6.38  2025-11-12 [1] CRAN (R 4.5.0)
#&gt;  evaluate      1.0.5   2025-08-27 [1] CRAN (R 4.5.0)
#&gt;  fastmap       1.2.0   2024-05-15 [1] CRAN (R 4.5.0)
#&gt;  HDDAData    * 1.0.1   2025-11-06 [1] Github (statOmics/HDDAData@b832c71)
#&gt;  htmltools     0.5.8.1 2024-04-04 [1] CRAN (R 4.5.0)
#&gt;  jquerylib     0.1.4   2021-04-26 [1] CRAN (R 4.5.0)
#&gt;  jsonlite      2.0.0   2025-03-27 [1] CRAN (R 4.5.0)
#&gt;  knitr         1.50    2025-03-16 [1] CRAN (R 4.5.0)
#&gt;  lifecycle     1.0.4   2023-11-07 [1] CRAN (R 4.5.0)
#&gt;  R6            2.6.1   2025-02-15 [1] CRAN (R 4.5.0)
#&gt;  rlang         1.1.6   2025-04-11 [1] CRAN (R 4.5.0)
#&gt;  rmarkdown     2.30    2025-09-28 [1] CRAN (R 4.5.0)
#&gt;  rstudioapi    0.17.1  2024-10-22 [1] CRAN (R 4.5.0)
#&gt;  sass          0.4.10  2025-04-11 [1] CRAN (R 4.5.0)
#&gt;  sessioninfo   1.2.3   2025-02-05 [1] CRAN (R 4.5.0)
#&gt;  xfun          0.54    2025-10-30 [1] CRAN (R 4.5.0)
#&gt;  yaml          2.3.10  2024-07-26 [1] CRAN (R 4.5.0)
#&gt; 
#&gt;  [1] /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/library
#&gt;  * ── Packages attached to the search path.
#&gt; 
#&gt; ──────────────────────────────────────────────────────────────────────────────</code></pre>
</details>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-alon1999broad" class="csl-entry">
Alon, Uri, Naama Barkai, Daniel A Notterman, Kurt Gish, Suzanne Ybarra, Daniel Mack, and Arnold J Levine. 1999. <span>“Broad Patterns of Gene Expression Revealed by Clustering Analysis of Tumor and Normal Colon Tissues Probed by Oligonucleotide Arrays.”</span> <em>Proceedings of the National Academy of Sciences</em> 96 (12): 6745–50.
</div>
</div>
</div>

<div id="rmd-source-code">---
title: "Lab 6: Large Scale Inference"
subtitle: "High Dimensional Data Analysis practicals"
author: "Adapted by Milan Malfait and Leo Fuhrhop"
date: "21 Feb 2022 <br/> (Last updated: 2025-11-26)"
references:
- id: alon1999broad
  type: article-journal
  author:
  - family: Alon
    given: Uri
  - family: Barkai
    given: Naama
  - family: Notterman
    given: Daniel A
  - family: Gish
    given: Kurt
  - family: Ybarra
    given: Suzanne
  - family: Mack
    given: Daniel
  - family: Levine
    given: Arnold J
  issued:
  - year: 1999
  title: Broad patterns of gene expression revealed by clustering analysis of tumor
    and normal colon tissues probed by oligonucleotide arrays
  container-title: Proceedings of the National Academy of Sciences
  publisher: National Acad Sciences
  page: 6745-6750
  volume: '96'
  issue: '12'
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  fig.width = 8,
  fig.asp = 0.618,
  out.width = "100%"
)
```

### [Change log](https://github.com/statOmics/HDDA/commits/master/Lab6-Large-Scale-Inference.Rmd) {-}

***

```{r libraries, warning=FALSE, message=FALSE}
## install packages with:
# if (!requireNamespace("remotes", quietly = TRUE)) {
#     install.packages("remotes")
# }
# remotes::install_github("statOmics/HDDAData")

library(HDDAData)
```

# Introduction

**In this lab session we will look at the following topics**

  - FWER
  - FDR
  - Multiple testing problem in a real dataset


## Testing many hypotheses {#testing-hypotheses}

To demonstrate the ideas we will be working with, we will simulate artificial data.
Note that since we are doing simulations, we can control everything and also know
exactly what the underlying "true" distribution is (sometimes also called the "ground truth").
However, keep in mind that this is an unlikely representation of real-world data.

In particular, we will simulate data for multiple hypothesis tests where
__the null hypothesis is always true__. This means that for the $i$-th test, if we let
$\mu_{1i}$ and $\mu_{2i}$ represent the means of the two populations of interest,
the null hypothesis for comparing the two means is
$H_{0i} : \mu_{1i} = \mu_{2i}$. Let the alternative hypothesis of this test be
$H_{1i} : \mu_{1i} \neq \mu_{2i}$, i.e., we perform a *two-tailed* test.
Suppose we have data collected from both populations, given by $X_{1i}$ and $X_{2i}$
of size $n_{1i}$ and $n_{2i}$, respectively, and assume that both populations
have the same known variance $\sigma_i^2$. Then we can test this hypothesis by using a
__Z-test statistic__, given by

$$ Z_i=\frac{\bar X_{1i}-\bar X_{2i}}{\sigma_i\sqrt{1/n_{1i}+1/n_{2i}}} \, .$$

Under the null hypothesis, the scores will be distributed as a *standard normal*
$Z_i\sim N(0,1)$. The null hypothesis is rejected in favor of the alternative if
$z_i < \phi_{\alpha_i/2}$ or $z_i > \phi_{1-\alpha_i/2}$, or equivalently if
$|z_i| > \phi_{1-\alpha_i/2}$, where $\phi_{\alpha_i/2}$ is the $\alpha_i/2$th
quantile of the standard normal distribution.

In a multiple hypothesis testing setting, we will perform $m$ tests using the
same test statistic. If the *null* were true for all hypotheses, we would end up
with a sample $z_1,\ldots,z_m$ from a standard normal distribution.

In this setting, we can only make one type of error: wrongly rejecting the null
hypothesis, i.e. a __type 1 error__. The probability of making this error is given
by

$$ \alpha_i=\text{P}\{\text{ reject } H_{0i} | H_{0i}\} $$

(the $| H_{0i}$ part should be read as "given that the null hypothesis is true").

If we now perform $m_0$ such tests (using the 0 subscript to denote that they
are all hypotheses for which the *null* is true), we can summarise the possible
outcomes as in the table below:

```{r, echo=FALSE}
knitr::kable(
  structure(list(
    `Accept H_0` = "U (True Negative)",
    `Reject H_0` = "V (False Positive)",
    `Total` = "m_0"
  ),
  .Names = c("Accept H_0", "Reject H_0", "Total"),
  row.names = c("H_0 True"),
  class = "data.frame"
  )
)
```


Here, $U$ and $V$ represent the total number of true negative and false positive
results we get, respectively, out of $m_0$ total tests.

As an example, let's simulate the scenario above. In the code below, the `replicate`
function is used to repeat the same procedure a number of times.
Essentially, the following steps are performed :

1. Sample $N = 50$ observations for 2 hypothetical groups from normal distributions
   with the same mean and known variance.
2. Calculate Z-test statistics based on the 2 group means.
3. Repeat steps 1 to 2 `m0` times.

This mimics performing `m0` hypothesis tests on data for which we know the null
hypothesis is true.

```{r}
# Set parameters for the simulation
N <- 50 # samples per group
m0 <- 1000 # number of hypothesis tests
mu_1 <- 3 # true mean group 1
mu_2 <- 3 # true mean group 2
sigma <- 1 # known variance
denom <- sigma * sqrt(2 / N) # denominator for Z-scores

set.seed(987654321) # seed for reproducibility
z_stats <- replicate(m0, {
  group1 <- rnorm(N, mean = mu_1, sd = sqrt(sigma))
  group2 <- rnorm(N, mean = mu_2, sd = sqrt(sigma))

  # Calculate Z-test statistic
  (mean(group2) - mean(group1)) / denom
})

# Visualize Z-test statistics
hist(z_stats, breaks = 50, freq = FALSE, main = "Z-test statistics under H_0")
# Overlay theoretical standard normal
lines(x <- seq(-5, 5, length.out = 100), dnorm(x), col = "dodgerblue", lwd = 3)
# Draw vertical lines at 2.5 and 97.5th percentiles
abline(v = qnorm(c(0.025, 0.975)), col = "firebrick", lty = 2, lwd = 3)
```

We see that the Z-test statistics nicely follow a standard normal distribution.
The vertical dashed lines indicate the 2.5th and 97.5th percentiles of the standard
normal. The regions outside these lines indicate the Z-scores that we would call
significant if we used a cut-off of $\alpha = 0.05$ for a *two-tailed* test.
So, even though we simulated data under the null hypothesis, our Z-test still
returns "significant" results for a number of cases just by chance!

Let's calculate the p-values for our hypothesis tests and see what the damage is.
To calculate the p-values, we use the `pnorm()` function in this case, which returns
the value of the standard normal CDF (i.e. `pnorm(x)` = $P(Z < x)$). Since we consider
a two-tailed test, we take the absolute values of the Z-scores and set the `lower.tail`
argument in `pnorm` to `FALSE` (by default it's `TRUE`), so that we get
`pnorm(abs(x), lower.tail = FALSE)` = $P(Z > |x|)$ and multiply this value by 2.

```{r}
null_pvals <- 2 * pnorm(abs(z_stats), lower.tail = FALSE)
alpha <- 0.05

hist(null_pvals, breaks = seq(0, 1, by = 0.05))
abline(v = alpha, col = "firebrick", lwd = 3)

called <- (null_pvals < alpha)

# V = number of false positives, in this case: all significant tests
(V <- sum(called))
mean(called) # V / m0
```

So we get 51 significant tests (false positives) out of a total of 1000, which is,
unsurprisingly, approximately equal to our significance cutoff $\alpha$.
Note also that the p-values are uniformly distibuted under the null hypothesis by definition.

If we had carried out only a few tests (say 10) it would be very unlikely to observe
a false positive (on average: 0.05 * 10 = 0.5 false positives) at $\alpha = 0.05$,
but since now we're carrying out so many, we're almost guaranteed to get false positives.
This is what is known as the __multiple hypothesis testing problem__.

Note that in real-world data the quantities $U$ and $V$ are *unknown* (because
we don't know the truth; if we did, we wouldn't have to carry out any hypothesis
tests in the first place). However, by using simulated data, we do know the truth
and so we can explore these quantities.


## The family-wise error rate

If we carry out many tests, we're almost guaranteed to get type I errors, just by
chance. Therefore the type I error rate is no longer a relevant metric.
Instead, we consider the __Familywise Error Rate (FWER)__, given by

$$
\text{FWER}=\text{P}\{V > 0\}
  = \text{P}\{\text{rejecting at least one } H_{0i}| H_0\}
$$

where $H_0$ is the intersection of all  partial nulls ($H_{0i}$) $i=1,\ldots,m_0$.
In general, we would prefer testing procedures that keep the FWER under control,
i.e. correct for the multiplicity problem.
So instead of choosing $\alpha$ to control the probability of getting a false positive
in each test, we try to control the FWER, i.e. the probability of getting
*at least* one false positive in our *set* of hypothesis tests.

In the following exercises, we will explore this idea on some synthetic data. We will
simulate performing hypothesis tests for which the null distribution is always true
for different values of `m0`.

To simplify the code, we will directly sample the Z-test statistics from
a standard normal (instead of sampling the individual groups and then performing
the test on them, as shown above). We can also skip the p-value calculation and
compare our Z-test statistics directly to the $(1 - \alpha) / 2$th quantile of the standard
normal distribution (using the `qnorm()` function).


### Tasks {-}

#### 1. Set `m0 = 5` and generate `m0` Z-test statistics by sampling from the standard normal distribution. Check which ones are significant using a cutoff `alpha = 0.05` {-}

#### 2. Calculate $V$, the number of significant tests. {-}

#### 3. Check whether `V > 0` (at least one significant) and store the result in a variable. {-}

#### 4. Repeat steps 1-3 1000 times and keep track of the result in step 3 by storing it in a vector. {-}

#### 5. Now compute the FWER as the proportion of times $V$ was greater than 0. {-}

#### 6. Repeat the same procedure for `m0 = 50` and `m0 = 1000`. Interpret the results. {-}

<details><summary>Solution</summary>

To avoid repeated code, the simulations over the different values of `m_0` are combined into one code block. 

```{r}
# Repeat 1000 times for m0 = 5, 50 and 1000
set.seed(987654321)
B <- 1000 # number of simulations
m0_vec <- c(5, 50, 1000)
alpha <- 0.05
# Initialize empty vector to store FWER results
fwer <- numeric(length = length(m0_vec))
names(fwer) <- paste("m0 =", m0_vec)

# Loop over m0 values and simulate test statistics
for (i in seq_along(m0_vec)) {
  # Simulate B times, each time recording V > 0
  at_least_one_signif <- replicate(B, {
    null_stats <- rnorm(m0_vec[i], mean = 0, sd = 1) # simulate z-test
    significant_tests <- abs(null_stats) > qnorm(1 - alpha / 2)
    sum(significant_tests) > 0
  })
  fwer[i] <- mean(at_least_one_signif) # proportion of V >= 1
}
fwer
```

</details>

# Controlling the FWER

We saw in the previous exercises that when conducting a large number of hypothesis
tests the FWER becomes unacceptably large. So what can we do to lower it?
The most straightforward way would be to just lower $\alpha$ (so we get less false positives).

Under the null hypothesis (so the p-values are uniformly distributed), we can write the FWER
for $m$ independent tests as

$$
\begin{align}
\text{FWER} = P(\text{at least one rejection}) &= 1 - P(\text{no rejections}) \\
  &= 1 - \prod_{i=1}^{m} P(p_i > \alpha) \\
  &= 1 - \prod_{i=1}^{m} (1 - \alpha) \\
  &= 1 - (1 - \alpha)^{m} \, ,
\end{align}
$$

where $p_i$ is the p-value for the $i$-th test.

Šidák's correction solves the above equation at a given FWER level (e.g., 0.05)
for $\alpha$ to control the FWER using 

$$ \alpha = 1 - (1 - \text{FWER})^{1/m} \, .$$

Without assuming independent tests, the Bonferroni correction uses that

$$
\begin{align}
\text{FWER} &= P(\text{at least one rejection}) \\
  &= P(\cup_{i=1}^m \{ p_i \le \alpha \}) \\
  &\le \sum_{i=1}^{m} P(p_i \le \alpha) \\
  &= m \alpha \, ,
\end{align}
$$

where the inequality follows from the union bound, and solves the equation
at a given FWER level (e.g., 0.05) for $\alpha$ to control the FWER with 

$$ \alpha = \frac{\text{FWER}}{m} \, .$$

### Tasks {-}

#### 1. Determine the `alpha` for Šidák's correction at `m0 = 5`, `m0 = 50` and `m0 = 1000`. Use these values to repeat the simulation from the previous exercise, and confirm that the FWER is controlled at 0.05. {-}

<details><summary>Solution</summary>

```{r}
# Repeat 1000 times for m0 = 5, 50 and 1000
set.seed(987654321)
B <- 1000 # number of simulations
m0_vec <- c(5, 50, 1000)
alpha_vec <- 1 - (1 - 0.05)^(1 / m0_vec)
alpha_vec
# Initialize empty vector to store FWER results
fwer <- numeric(length = length(m0_vec))
names(fwer) <- paste("m0 =", m0_vec)

# Loop over m0 values and simulate test statistics
for (i in seq_along(m0_vec)) {
  # Simulate B times, each time recording V > 0
  at_least_one_signif <- replicate(B, {
    null_stats <- rnorm(m0_vec[i], mean = 0, sd = 1) # simulate z-test
    significant_tests <- abs(null_stats) > qnorm(1 - alpha_vec[i] / 2)
    sum(significant_tests) > 0
  })
  fwer[i] <- mean(at_least_one_signif) # proportion of V >= 1
}
fwer
```

</details>

#### 2. Determine the `alpha` for the Bonferroni correction and repeat the simulation one more time. How do the results compare to Šidák's correction? {-}

<details><summary>Solution</summary>

```{r}
# Repeat 1000 times for m0 = 5, 50 and 1000
set.seed(987654321)
B <- 1000 # number of simulations
m0_vec <- c(5, 50, 1000)
alpha_vec <- 0.05 / m0_vec
alpha_vec
# Initialize empty vector to store FWER results
fwer <- numeric(length = length(m0_vec))
names(fwer) <- paste("m0 =", m0_vec)

# Loop over m0 values and simulate test statistics
for (i in seq_along(m0_vec)) {
  # Simulate B times, each time recording V > 0
  at_least_one_signif <- replicate(B, {
    null_stats <- rnorm(m0_vec[i], mean = 0, sd = 1) # simulate z-test
    significant_tests <- abs(null_stats) > qnorm(1 - alpha_vec[i] / 2)
    sum(significant_tests) > 0
  })
  fwer[i] <- mean(at_least_one_signif) # proportion of V >= 1
}
fwer
```

</details>

In general, controlling the FWER at levels such as 0.05 with Šidák's correction
or the Bonferroni correction is very conservative, i.e., if there were true alternatives
(which we did not consider in the simulations), we might miss them.


# The False Discovery Rate (FDR)

Now let's consider a situation where we have both true null and true alternative
cases. In this situation we are prone to two types of errors: false positives
(type 1, as in the previous case) and false negatives (type 2, i.e. accepting
$H_0$ while it's actually false).

We can then extend the table from [before](#testing-hypotheses):

```{r, echo=FALSE}
knitr::kable(
  structure(list(
    `Accept H_0` = c("U (True Negative)", "T (False Negative)", "W"),
    `Reject H_0` = c("V (False Positive)", "S (True Positive)", "R"),
    `Total` = c("m_0", "m_1", "m")
  ),
  .Names = c("Accept H_0", "Reject H_0", "Total"),
  row.names = c("H_0 True", "H_1 True", "Total"),
  class = "data.frame"
  )
)
```

In addition to the FWER, another important quantity in multiple hypothesis
testing is the __False Discovery Rate (FDR)__:

$$
\text{FDR} = E[\frac{V}{R}] \, ,
$$

which is the expected proportion of false positives among all positive tests.
The ratio $V/R$ is also called the False Discovery Proportion (FDP).

However, the only quantities from the table above that are observable
in real data are $W$, $R$ and $m$. So to illustrate the concept, we will again
make use of simulated data.

In the code below, 1.000 tests are simulated, for which 90% come from cases
where the null is true and the other 10% for which the alternative is true.
We assume the *effect size* (difference between the 2 groups) to be equal to 3
under the alternative, so that we can sample the test statistics from a normal
distribution with mean 3.

```{r}
m <- 1000 # total number of hypotheses
p0 <- 0.9 # 90% of cases are true null
m0 <- round(p0 * m)
m1 <- round((1 - p0) * m)

set.seed(987654321)
alpha <- 0.05
null_stats <- rnorm(m0)
alt_stats <- rnorm(m1, mean = 3)
null_tests <- abs(null_stats) > qnorm(1 - alpha / 2)
alt_tests <- abs(alt_stats) > qnorm(1 - alpha / 2)

(U <- sum(!null_tests)) # true negatives
(V <- sum(null_tests)) # false positives
(S <- sum(alt_tests)) # true positives
(T <- sum(!alt_tests)) # false negatives

R <- V + S # total number of positives

(FDP <- V / R)
```

We see that 35.38% of the positive cases are actually false positives.
To get an idea of the FDR, the expected FDP, we will repeat this procedure a
number of times in the following exercises.

### Tasks {-}

#### 1. Set `m0 = 90` and `m1 = 10` and generate `m = 100` test statistics, `m0` of them from the standard normal distribution $N(0, 1)$ (null distribution) and m1 of them from $N(3, 1)$ (the alternative distribution). {-}


#### 2. Test for significance at `alpha = 0.05` by comparing the test statistics to the standard normal, for both sets of tests. {-}


#### 3. Calculate the FDP. {-}


#### 4. Repeat steps 1-3 1000 times and keep track of the FDP for each iteration. Then estimate the FDR as the mean of the FDPs. Interpret the results. {-}

<details><summary>Solution</summary>

```{r}
set.seed(987654321)
B <- 1000 # number of simulations
m0 <- 90
m1 <- 10
alpha <- 0.05

# Simulate B times, each time recording V > 0
FDP <- replicate(B, {
  null_stats <- rnorm(m0, mean = 0)
  alt_stats <- rnorm(m1, mean = 3)
  null_tests <- abs(null_stats) > qnorm(1 - alpha / 2)
  alt_tests <- abs(alt_stats) > qnorm(1 - alpha / 2)
  V <- sum(null_tests) # false positives
  S <- sum(alt_tests) # true positives
  R <- V + S
  # Return V / R
  if (R == 0) 0 else V / R
})

(FDR <- mean(FDP))
```

The FDR tells us the expected proportion of positive tests that will be false
positives.

</details>

As with the FWER, we can now think of methods to control the FDR. A very popular
method that is widely used in large-scale inference for high-throughput
sequencing data is the __Benjamini-Hochberg correction__. It *adjusts* the
p-values such that the FDR is controlled at a certain level. Note that this can
equivalently be formulated as adjusting the significance cut-off $\alpha$, similarly
to how we defined the FWER-controlling techniques.

### Tasks {-}

#### 1. Adjust the simulation code from the previous task to estimate the FDR under the Benjamini-Hochberg correction. {-}
Hint: Calculate p-values of the simulated null and alternative test statistics. Concatenate all p-values into a single vector and apply the Benjamini-Hochberg correction using `p.adjust` with `method = "BH"`. Compare the resulting adjusted p-values to $\alpha$ to estimate the FDP and FDR.

<details><summary>Solution</summary>

```{r}
set.seed(987654321)
B <- 1000 # number of simulations
m0 <- 90
m1 <- 10
alpha <- 0.05

# Simulate B times, each time recording V > 0
FDP <- replicate(B, {
  null_stats <- rnorm(m0, mean = 0)
  alt_stats <- rnorm(m1, mean = 3)
  p_null <- (2 * pnorm(-abs(null_stats)))
  p_alt <- (2 * pnorm(-abs(alt_stats)))
  p_adj <- p.adjust(c(p_null, p_alt), method = "BH")
  V <- sum(p_adj[1:m0] <= alpha) # false positives
  S <- sum(p_adj[(m0 + 1):(m0 + m1)] <= alpha) # true positives
  R <- V + S
  # Return V / R
  if (R == 0) 0 else V / R
})

(FDR <- mean(FDP))
```

</details>

#### 2. Repeat the simulation using `method = "bonferroni"`. Compare the resulting FDR and interpret. {-}
Note: The Bonferroni correction can either be formulated as adjusting the significance cut-off to $\alpha / m$ or, equivalently, as a p-value correction that multiplies each p-value by $m$. 

<details><summary>Solution</summary>

```{r}
set.seed(987654321)
B <- 1000 # number of simulations
m0 <- 90
m1 <- 10
alpha <- 0.05

# Simulate B times, each time recording V > 0
FDP <- replicate(B, {
  null_stats <- rnorm(m0, mean = 0)
  alt_stats <- rnorm(m1, mean = 3)
  p_null <- (2 * pnorm(-abs(null_stats)))
  p_alt <- (2 * pnorm(-abs(alt_stats)))
  p_adj <- p.adjust(c(p_null, p_alt), method = "bonferroni")
  V <- sum(p_adj[1:m0] <= alpha) # false positives
  S <- sum(p_adj[(m0 + 1):(m0 + m1)] <= alpha) # true positives
  R <- V + S
  # Return V / R
  if (R == 0) 0 else V / R
})

(FDR <- mean(FDP))
```

__Interpretation__: Both the Bonferroni correction and the Benjamini-Hochberg correction control the FDR at the desired level $\alpha$.
However, the Bonferroni correction is more conservative, resulting in an FDR that is much lower than $\alpha$. This means that the Bonferroni method results in a larger number of false negatives and could possibly lead to us missing interesting discoveries.

</details>

# Multiple testing on the data by Alon *et al.* (1999)

We will take another look at the dataset by @alon1999broad on gene expression
levels in 40 tumour and 22 normal colon tissue samples.
We previously used this data in [Lab 4](./Lab4-Sparse-PCA-LDA.html).
This time, we are interested in finding genes that are expressed *significantly*
different between the tumor and normal samples. In other words, we will perform
hypothesis tests for each of the __2000 genes__. This is clearly a multiple
testing problem.

```{r load-data}
data("Alon1999")
str(Alon1999[, 1:10]) # first 10 columns
table(Alon1999$Y)  # contingency table of tissue types
```

As a reminder, the data consists of gene expression levels in 40 tumor and 22
normal colon tissue samples. The expression on 6500 human genes were measured
using the Affymetrix oligonucleotide array.
As in @alon1999broad, we use the *2000 genes with the highest minimal intensity*
across the samples.

The dataset contains one variable named `Y` with the values `t` and `n`.
This variable indicates whether the sample came from tumourous (`t`) or
normal (`n`) tissue.

### Tasks {-}

#### 1. Perform two-sample t-tests for each gene, comparing the tumor and normal tissue groups. Store the p-values and test statistics. {-}

Hint: The `t.test()` function computes a two-sample t-test with *unequal variances* ([Welch's t-test][Welch]) by default. The return object contains the elements `p.value` and `statistic`.

<details><summary>Solution</summary>

```{r}
# Split gene data and grouping variable and convert to matrix
gene_data <- as.matrix(Alon1999[, -1])
group <- Alon1999$Y

# Use `apply` to perform t-tests over all columns
t_test_results <- t(apply(gene_data, 2, function(x) {
  t_test <- t.test(x ~ group)
  p_val <- t_test$p.value
  stat <- t_test$statistic
  c(stat, "p_val" = p_val)
}))

head(t_test_results)
```

</details>

#### 2. Plot a histogram of the p-values and interpret. {-}

<details><summary>Solution</summary>

Setting `breaks` from 0 to 1 with a width of 0.05 makes sense for p-values because
the first bar will then represent the number of p-values smaller than 0.05 (which
we often use as cutoff).

```{r p_val-hist}
p_vals <- t_test_results[, "p_val"]
hist(
  p_vals,
  breaks = seq(0, 1, by = 0.05), main = "", xlab = "p-value",
  ylim = c(0, 500)
)
```

__Interpretation__: The histogram appears close to a uniform distribution for the larger p-values, but has a lot more small p-values than would be expected under a uniform distribution.
Recall that if the null hypothesis (the gene expression is the same for tumor and normal tissue) were to hold for all genes, the p-values would be uniformly distributed on $[0, 1]$.

</details>

#### 3. How many discoveries / significant genes do you find when using $\alpha = 0.05$? {-}

<details><summary>Solution</summary>

```{r}
alpha <- 0.05
sum(p_vals <= alpha)
```

</details>

#### 4. Control the FWER at 5% using Šidák's correction. How many genes are significant after the correction? {-}

<details><summary>Solution</summary>

```{r}
m <- nrow(t_test_results)
fwer <- 0.05
adj_alpha <- 1 - (1 - fwer)^(1 / m)

# Number of significant discoveries
sum(p_vals <= adj_alpha)
```

By rearranging the equation $$ \alpha = 1 - (1 - \text{FWER})^{1/m} \, ,$$
we can equivalently adjust the p-values via
$$1 - (1 - p)^m$$
and compare them to the cut-off 5\%:

```{r}
sidak_p_vals <- 1 - (1 - p_vals)^m

# Number of significant discoveries
sum(sidak_p_vals <= fwer)
```

</details>

#### 5. Control the FWER at 5% using the Bonferroni correction. How many genes are significant after the correction? {-}

<details><summary>Solution</summary>

By adjusting $\alpha$:

```{r}
adj_alpha <- fwer / m

# Number of significant discoveries
sum(p_vals <= adj_alpha)
```

Or equivalently, by adjusting the p-values:

```{r}
bonferroni_p_vals <- p.adjust(p_vals, method = "bonferroni")

# Number of significant discoveries
sum(bonferroni_p_vals <= fwer)
```

</details>

#### 6. Control the FDR at 5% using the Benjamini-Hochberg method. How many genes are significant after the correction? {-}

<details><summary>Solution</summary>

```{r}
bh_p_vals <- p.adjust(p_vals, method = "BH")

sum(bh_p_vals < 0.05)
```

</details>

#### 7. Plot the adjusted p-values from the Benjamini-Hochberg method against the original p-values. What is the effect of the adjustment? {-}
Bonus: Compute the adjusted p-values based on Šidák's correction and the Bonferroni correction. Add them to the plot and compare.

<details><summary>Solution</summary>

Visualize the adjusted p-values for all three methods in the same figure:

```{r fdr-adjust}
cols <- c("#9d05f4", "#ecd033", "#01d9d9")

plot(
  p_vals[order(p_vals)], bh_p_vals[order(p_vals)],
  pch = 19, cex = 0.6, xlab = "p-value",
  ylab = "Adjusted p-value", col = cols[1]
)
points(
  p_vals[order(p_vals)], bonferroni_p_vals[order(p_vals)],
  pch = 19, cex = 0.6, col = cols[2]
)
points(
  p_vals[order(p_vals)], sidak_p_vals[order(p_vals)],
  pch = 19, cex = 0.6, col = cols[3]
)
abline(a = 0, b = 1)
legend("bottomright",
  c("Benjamini-Hochberg", "Bonferroni", "Sidak"),
  col = cols, pch = 19
)
```

__Interpretation__: All three methods adjust the p-values upwards to account for multiple testing. The adjustment is more pronounced for small p-values, while for higher p-values, the curves of the adjusted p-values flatten out. 
The Benjamini-Hochberg method at a FDR of 5% is less conservative compared to the other two methods. Both Šidák's correction and the Bonferroni correction at an FWER of 5% adjust the majority of p-values to exactly 1 and only a small proportion of p-values that are originally close to zero remains smaller than 1.

</details>

# Further reading {-}

- Chapter 6 of the [Biomedical Data Science](http://genomicsclass.github.io/book/) book

[Welch]:https://en.wikipedia.org/wiki/Welch%27s_t-test

```{r, child="_session-info.Rmd"}
```

# References {-}
</div>
<div class="footer">
    <hr>
    This work is licensed under the <a href= "https://creativecommons.org/licenses/by-nc-sa/4.0">
    CC BY-NC-SA 4.0</a> licence.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Lab6-Large-Scale-Inference.Rmd");
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>


</body>
</html>
